raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -937,17 +938,22 @@ private void maybeSendOrHandleAppendRequest(long currentTimeMs) {
         if (unsentAppend == null || unsentAppend.isCancelled())
             return;
 
-        if (unsentAppend.isTimedOut(currentTimeMs)) {
-            unsentAppend.fail(new TimeoutException());
-        } else if (quorum.isLeader()) {
-            LeaderState leaderState = quorum.leaderStateOrThrow();
-            int epoch = quorum.epoch();
-            OptionalLong baseOffsetOpt = maybeAppendAsLeader(leaderState, unsentAppend.records);
-            if (baseOffsetOpt.isPresent()) {
-                unsentAppend.complete(new OffsetAndEpoch(baseOffsetOpt.getAsLong(), epoch));
+        if (quorum.isLeader()) {
+            if (unsentAppend.isTimedOut(currentTimeMs)) {
+                unsentAppend.fail(new TimeoutException());
             } else {
-                unsentAppend.fail(new InvalidRequestException("Leader refused the append"));
+                LeaderState leaderState = quorum.leaderStateOrThrow();
+                int epoch = quorum.epoch();
+                OptionalLong baseOffsetOpt = maybeAppendAsLeader(leaderState, unsentAppend.records);
+                if (baseOffsetOpt.isPresent()) {
+                    unsentAppend.complete(new OffsetAndEpoch(baseOffsetOpt.getAsLong(), epoch));
+                } else {
+                    unsentAppend.fail(new InvalidRequestException("Leader refused the append"));
+                }
             }
+        } else {
+            unsentAppend.fail(new NotLeaderForPartitionException("Append refused since this node is no longer " +
```

Comment:
This was a bug fix, probably the result of removing the AppendRecord API which would previously be forwarded to the leader. Now if we are not leader, we just reject the append.